Source/WTF/wtf/text/StringCommon.h

```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
+            return pointer + index;
+    }
+    if (runway == length)
+        return nullptr;
 
-ALWAYS_INLINE const uint16_t* memchr16(const uint16_t* pointer, uint16_t character, size_t length)
+    ASSERT(index < length);
+    return static_cast<const uint8_t*>(memchr(pointer + index, character, length - index));
+}
+
+#if CPU(ARM64)
+WTF_EXPORT_PRIVATE const uint16_t* find16AlignedImpl(const uint16_t* pointer, uint16_t character, size_t length);
+
+ALWAYS_INLINE const uint16_t* find16(const uint16_t* pointer, uint16_t character, size_t length)
 {
     // We take `size_t` length instead of `unsigned` because,
     // 1. It is aligned to memchr.
-    // 2. It allows us to use memchr16 for 4GB~ vectors, which can be used in JSC ArrayBuffer (4GB wasm memory).
+    // 2. It allows us to use find16 for 4GB~ vectors, which can be used in JSC ArrayBuffer (4GB wasm memory).
 
     // If the pointer is unaligned to 16bit access, then SIMD implementation does not work. But ARM64 allows unaligned access.
     // Fallback to a simple implementation. We also use it for smaller memory where length is less than 16.
-    constexpr unsigned memchrThresholdLength = 32;
-    static_assert(!(memchrThresholdLength % 8), "It should be 16byte aligned to make memchr16AlignedImpl simpler");
+    constexpr size_t thresholdLength = 32;
+    static_assert(!(thresholdLength % (16 / sizeof(uint16_t))), "It should be 16byte aligned to make find16AlignedImpl simpler");
 
     // For first check `threshold - (unaligned >> 1)` characters, we use normal loop.
     // This can (1) align pointer to 16byte size so that SIMD loop gets simpler and (2) handle cases
     // having a character in the beginning of the string efficiently.
     uintptr_t unaligned = (reinterpret_cast<uintptr_t>(pointer) & 0xf);
 
     size_t index = 0;
-    size_t runway = std::min<size_t>(memchrThresholdLength - (unaligned >> 1), length);
+    size_t runway = std::min<size_t>(thresholdLength - (unaligned / sizeof(uint16_t)), length);
+    for (; index < runway; ++index) {
+        if (pointer[index] == character)
+            return pointer + index;
+    }
+    if (runway == length)
+        return nullptr;
+
+    ASSERT(index < length);
+    return find16AlignedImpl(pointer + index, character, length - index);
+}
+#else
+ALWAYS_INLINE const uint16_t* find16(const uint16_t* pointer, uint16_t character, size_t length)
+{
+    for (size_t index = 0; index < length; ++index) {
+        if (pointer[index] == character)
+            return pointer + index;
+    }
+    return nullptr;
+}
+#endif
+
+#if CPU(ARM64)
+WTF_EXPORT_PRIVATE const uint32_t* find32AlignedImpl(const uint32_t* pointer, uint32_t character, size_t length);
+
+ALWAYS_INLINE const uint32_t* find32(const uint32_t* pointer, uint32_t character, size_t length)
+{
+    constexpr size_t thresholdLength = 32;
+    static_assert(!(thresholdLength % (16 / sizeof(uint32_t))), "it should be 16byte aligned to make find32AlignedImpl simpler");
```

Comment:
> I think "it should be" is not clear. I would say "length threshold should be". I would write "16-byte aligned" instead of "16byte aligned".

Fixed.

> I would also like a comment explaining why the threshold should not be even higher. Did we measure it? It would be cool to leave a comment behind saying why we chose it so that someone later could choose an even better one, but has a head start understanding why we set it to 32 now.

Yeah, this is picked from measurement to keep the smaller data case still the sane.

> Same thoughts on the similar code we already have in the 16-bit version above.